# Python Day 1

## 1. Python Environment

### 1.1 Install

> Clicking on a hyperlink, the corresponding target page will be opened.

- [Python](https://www.python.org/downloads/mac-osx/)
- [Anaconda](https://www.anaconda.com/products/individual#macos)
    - Highly recommended for packages and environments management
    - Make different python versions coexist sucessefully in one system

#### Some common Anoconda commands:
* `conda info --envs` : lists all environments
* `conda activate <env name>`: activate an environment
* `conda deactivate`: deactivate the current environment
* `conda list` : list all packages installed
* `conda create --name <env name> python=3 astroid babel` : create new environment, specify version of python, and install packages
* `conda install <package name>` install a pachage
* `conda env remove -n <env name> ` : delete environment

### 1.2 Programming

- [Pycharm](https://www.jetbrains.com/pycharm/download/#section=windows)
    - Community version is enough
    - Highly recommended IDE (Integrated Develepment Environment) with GUI (Graphical User Interface)
- [Jupyter notebook](https://jupyter.org/)
    - Interactive、Readable、Shareable、Supporting many languages、Useful in a server

# 2. Python Introduction

## 2.1 Python Data Types

Python has rich built-in data types. Here are some tutorials on Python Data Types for your reference.
- [W3schools](https://www.w3schools.com/python/python_datatypes.asp)
- [TutorialsTeacher](https://www.tutorialsteacher.com/python/python-data-types)

Some basic date types are as followed. 

- int

In [37]:
a = 3
type(a)

int

- float

In [38]:
b = 3.1415
type(b)

float

- string

In [39]:
c = 'c'
d = 'iamd'
print(type(c))
print(type(d))

<class 'str'>
<class 'str'>


- list

In [40]:
mylist = [1,2,3, 4, 5]
def testAPI():
    mylist.pop()
    mylist.append(10)
testAPI()
print(mylist)

[1, 2, 3, 4, 10]


## 2.2 Function

- A block for completing a specific task.
- It will be executed when called.
- Inputs: some data (parameters) that are passed into it for computation. 
- Output: the computed result, passed out by `return`.

In [41]:
# Calculate the sum of two numbers
def sumOfTwoNumbers(x, y):
    return x+y

# Calculate the sum of all elements inside a list
def sumOfListElements(alist):
    result = 0
    for element in alist:
        result = result + element
    return result

result = sumOfTwoNumbers(100, 977832)
print("The first result is {}.".format(result))

result = sumOfListElements([7, 7, 8])
print("The second result is {}.".format(result))

The first result is 977932.
The second result is 22.


# 3. Python Examples

## 3.1 Factorial Calculation

This example calculates the factorial of an integer. Two ways are given to complete this task, though with different efficiency.

- The first way

$$ f(n) = n! = 1\times2\times3\times...\times n .$$

In [42]:
# Iterative manner
def iterativeFactorial(n):
    result = 1
    for i in range(1, n+1):
        result = result * i
    return result

- The second way

$$ f(n) = \left\{
\begin{array}{l}
n \times f(n) & (n>1) \\
1  & (n=1).\\
\end{array}
\right.
$$

In [43]:
# Recursive manner
def recursiveFactorial(n):
    if n > 1:
        return n * recursiveFactorial(n-1)
    else:
        return 1

In [44]:
n = 10
print("The iterative factorial result is {}.".format(iterativeFactorial(n)))
print("The recursive factorial result is {}.".format(recursiveFactorial(n)))

The iterative factorial result is 3628800.
The recursive factorial result is 3628800.


## 3.2 To Find $\pi$

The below formula for computing $\pi$ is from [here](https://www.craig-wood.com/nick/articles/pi-archimedes/).

The Perimeter of a $n$-sides regular polygon is $d_n * n$ with $d_n$ as the length of one edge. Meanwhile, the Diameter of the polygon's mimimum circumcircle is 2.

Thus, with $n$ become larger and larger, the differences between $(d_n * n) / 2$ and $\pi$ are constantly decreasing.

$$d_{2n} = \sqrt{2 - 2\sqrt{1-\frac{d_n^2}{4}}}.$$

$$\pi \approx \frac{d_n * n}{2}, \text{with } n \text{ approaching to } \infty.$$

In [45]:
from math import pi
from math import sqrt

def computePi(iters):
    sides = 4
    dn = sqrt(2)
    print("Iters | Sides | Results | Error")
    print("------|-------|---------|------")
    for i in range(iters):
        result = dn * sides / 2
        print("%2d | %8d | %.10f | %.10f" % (i, sides, result, pi - result))
        sides = sides * 2
        dn = sqrt(2 - 2 * sqrt(1 - dn*dn/4) )

computePi(16)

Iters | Sides | Results | Error
------|-------|---------|------
 0 |        4 | 2.8284271247 | 0.3131655288
 1 |        8 | 3.0614674589 | 0.0801251947
 2 |       16 | 3.1214451523 | 0.0201475013
 3 |       32 | 3.1365484905 | 0.0050441630
 4 |       64 | 3.1403311570 | 0.0012614966
 5 |      128 | 3.1412772509 | 0.0003154027
 6 |      256 | 3.1415138011 | 0.0000788524
 7 |      512 | 3.1415729404 | 0.0000197132
 8 |     1024 | 3.1415877253 | 0.0000049283
 9 |     2048 | 3.1415914215 | 0.0000012321
10 |     4096 | 3.1415923456 | 0.0000003080
11 |     8192 | 3.1415925765 | 0.0000000770
12 |    16384 | 3.1415926335 | 0.0000000201
13 |    32768 | 3.1415926548 | -0.0000000012
14 |    65536 | 3.1415926453 | 0.0000000083
15 |   131072 | 3.1415926074 | 0.0000000462


To see the results in a markdown table as blow.

Iters | Sides | Results | Error
------|-------|---------|------
 0 |        4 | 2.8284271247 | 0.3131655288
 1 |        8 | 3.0614674589 | 0.0801251947
 2 |       16 | 3.1214451523 | 0.0201475013
 3 |       32 | 3.1365484905 | 0.0050441630
 4 |       64 | 3.1403311570 | 0.0012614966
 5 |      128 | 3.1412772509 | 0.0003154027
 6 |      256 | 3.1415138011 | 0.0000788524
 7 |      512 | 3.1415729404 | 0.0000197132
 8 |     1024 | 3.1415877253 | 0.0000049283
 9 |     2048 | 3.1415914215 | 0.0000012321
10 |     4096 | 3.1415923456 | 0.0000003080
11 |     8192 | 3.1415925765 | 0.0000000770
12 |    16384 | 3.1415926335 | 0.0000000201
13 |    32768 | 3.1415926548 | -0.0000000012
14 |    65536 | 3.1415926453 | 0.0000000083
15 |   131072 | 3.1415926074 | 0.0000000462

In [46]:
from math import sqrt

type(sqrt(2))

float

You can find that errors are increasing after 13 iterations. This is because that ** python float type only has limited precision** (about 17 digits). In fact, this is a problem in almost all programming languages and has great influence in many numerical computing tasks. Here, we can use `Decimal` class in python to avoid this problem in this scenario.

In [47]:
from decimal import Decimal, getcontext

def computePi(iters):
    sides = 4
    dn_squared = Decimal(2)
    print("Iters | Sides | Results")
    print("------|-------|--------")
    for i in range(iters):
        result = dn_squared.sqrt() * int(sides / 2)
        print("%2d | %20d | %s " % (i, sides, result))
        sides = sides * 2
        dn_squared = 2 - 2 * (1 - dn_squared/4).sqrt()

getcontext().prec = 50 # 50 digits of precision
computePi(30)

Iters | Sides | Results
------|-------|--------
 0 |                    4 | 2.8284271247461900976033774484193961571393437507538 
 1 |                    8 | 3.0614674589207181738276798722431909340907564998852 
 2 |                   16 | 3.1214451522580522855725578956323558548430658840311 
 3 |                   32 | 3.1365484905459392638142580444365390675563735413589 
 4 |                   64 | 3.1403311569547529123171185243316901321437032336505 
 5 |                  128 | 3.1412772509327728680620197707882144083796632626652 
 6 |                  256 | 3.1415138011443010763285150594568223079353138153866 
 7 |                  512 | 3.1415729403670913841358001102707614295336377938332 
 8 |                 1024 | 3.1415877252771597006288542627019187393992808573309 
 9 |                 2048 | 3.1415914215111999739979717637408339557475626459655 
10 |                 4096 | 3.1415923455701177423403759941573699303052060957299 
11 |                 8192 | 3.1415925765848726656816060922378

# 3. Learning Resources

- [Python offical tutorial](https://docs.python.org/3/tutorial/index.html)
- [Python in w3school](https://www.w3schools.com/python/python_datatypes.asp)
- You can search for other ones you like just with google...

- [Markdown](https://www.markdownguide.org/basic-syntax/)